### PR TITLE
CRD server wait for CRDs to sync before reporting healthy

### PR DESCRIPTION
**Pull request by deads2k**
A kube-apiserver can report healthy before it is able to serve CRD requests for all the CRDs already registered.  We cannot eliminate the race for new types being added after a server is up, but we can be sure we have a view of the world that is consistent with a particular resourceVersion.  This pull adds a health check to do that

/kind bug
/priority important-soon

```release-note 
NONE
```

@kubernetes/sig-api-machinery-bugs

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go [hunk 1]
```diff
@@ -210,6 +211,18 @@ func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget)
 		return nil
 	})
 
+	// we don't want to report healthy until we can handle all CRDs that have already been registered.  Waiting for the informer
+	// to sync makes sure that the lister will be valid before we begin.  There may still be races for CRDs added after startup,
```
Comment (lavalamp) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go]: This is true for readiness check, but not for liveness check. If we aren't healthy, kubelet will restart us. Is that what we want?

**Comment by lavalamp** [comment]
/hold

(Just so Han has a chance to see it)

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go (hunk 1 above)
Comment (deads2k) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go]: > This is true for readiness check, but not for liveness check. If we aren't healthy, kubelet will restart us. Is that what we want?

`HasSynced` is a "past the post" function that never returns to a false state.

**Comment by lavalamp** [comment]
Yes, but if it takes a long time to finish?

**Review by logicalhan — COMMENTED** [review]
Out of curiosity, is there any reason why you did not opt to wrap this in a poststarthook? The logic in this health check seems to more conform with the sort of health check generated by poststarthooks (i.e. "past the post" as you describe it).

**Comment by deads2k** [comment]
> Yes, but if it takes a long time to finish?

I think it's still better to report unhealthy.  If you report healthy, you get included in the service.  If you're included in the service, then traffic can be routed to you.  If traffic is routed to you and you 404 on a GET of a CRD, you can end up facing pretty negative consequences.  Imagine it's an existence check from the garbage collector and we report a 404 back, so GC has confirmation the resource is gone.

**Comment by deads2k** [comment]
> Out of curiosity, is there any reason why you did not opt to wrap this in a poststarthook? The logic in this health check seems to more conform with the sort of health check generated by poststarthooks (i.e. "past the post" as you describe it).

I don't have any real factorization preference.

**Review by logicalhan — APPROVED** [review]
/lgtm (though I have a tiny preference for wrapping it in a poststarthook)

**Comment by mbohlool** [comment]
/lgtm

nice to at least be consistent when API server starts up. This should work good in HA clusters too. Thanks for doing this @deads2k

**Comment by lavalamp** [comment]
> I think it's still better to report unhealthy. If you report healthy, you get included in the service.

@deads2k You are talking about readiness. I agree it is not ready. I am talking about liveness; I do not agree it's not alive.

Unfortunately, these are conflated in apiserver. It's a real :man_facepalming: that we don't do this right with our own components.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go [hunk 2]
```diff
@@ -209,6 +210,14 @@ func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget)
 		go finalizingController.Run(5, context.StopCh)
 		return nil
 	})
+	// we don't want to report healthy until we can handle all CRDs that have already been registered.  Waiting for the informer
+	// to sync makes sure that the lister will be valid before we begin.  There may still be races for CRDs added after startup,
+	// but we won't go healthy until we can handle the ones already present.
+	s.GenericAPIServer.AddPostStartHookOrDie("crd-informer-synced", func(context genericapiserver.PostStartHookContext) error {
```
Comment (logicalhan) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go]: Oh sweet! Thanks for doing that! 😃

**Review by logicalhan — APPROVED** [review]
/lgtm

Really appreciate you making the change!

**Pull request merged**